File: conversoes/afn_para_afd.py
from typing import Dict, Any, Set, List
from af_utils import mover, normalizar_nome_estado, estados_finais_compostos

def converter(automato: Dict[str, Any]) -> Dict[str, Any]:
    """
    Converte um AFN em AFD usando subconjuntos e RENOMEIA os estados
    para um formato sequencial simples (q0, q1, q2...) ou amigável.
    """
    alfabeto = automato.get("alfabeto", [])
    estados_afn = set(automato.get("estados", []))
    estados_finais_afn = set(automato.get("estados_finais", []))

    estado_inicial_afn = set(automato.get("estados_iniciais", []))
    nome_inicial_bruto = normalizar_nome_estado(estado_inicial_afn)

    # Listas temporárias com os nomes "feios" (ex: "{q0,q1}")
    estados_brutos = [nome_inicial_bruto]
    transicoes_brutas = []
    
    fila = [estado_inicial_afn]
    visitados = [estado_inicial_afn]

    # --- 1. Lógica dos Subconjuntos (Gera o AFD Matemático) ---
    while fila:
        estado_atual = fila.pop(0)
        nome_atual = normalizar_nome_estado(estado_atual)

        for simbolo in alfabeto:
            # Calcula destino (mesmo vazio)
            destino = mover(automato, estado_atual, simbolo)
            nome_destino = normalizar_nome_estado(destino)
            
            transicoes_brutas.append([nome_atual, nome_destino, simbolo])

            # Se é um estado novo, adiciona na fila
            if destino not in visitados:
                visitados.append(destino)
                fila.append(destino)
                estados_brutos.append(nome_destino)

    # --- 2. Etapa de Renomeação (Estética) ---
    # Vamos criar um mapa: "{q0,q1}" -> "q1" (ou S1, etc)
    mapa_nomes = {}
    novos_estados = []
    novos_iniciais = []
    novos_finais = []
    
    # O estado inicial será sempre o primeiro da lista, vamos chamá-lo de S0 ou q0
    # Dica: Usar "S" (State) evita confusão com os "q" do original, mas você pode usar "q"
    PREFIXO = "S" 

    for i, nome_bruto in enumerate(estados_brutos):
        # Se for o estado de erro {}, podemos dar um nome especial
        if nome_bruto == "{}":
            novo_nome = "q_erro" # Ou "limbo", ou ""
        else:
            novo_nome = f"{PREFIXO}{i}"
        
        mapa_nomes[nome_bruto] = novo_nome
        novos_estados.append(novo_nome)

    # Recriar transições com os novos nomes
    novas_transicoes = []
    for origem, destino, simbolo in transicoes_brutas:
        novas_transicoes.append([
            mapa_nomes[origem], 
            mapa_nomes[destino], 
            simbolo
        ])

    # Definir iniciais e finais com os novos nomes
    novos_iniciais = [mapa_nomes[nome_inicial_bruto]]

    # Para finais, precisamos checar o conjunto original em 'visitados'
    # A lista 'visitados' está na mesma ordem de 'estados_brutos'
    for i, conjunto_original in enumerate(visitados):
        if estados_finais_compostos(conjunto_original, estados_finais_afn):
            nome_novo = mapa_nomes[estados_brutos[i]]
            novos_finais.append(nome_novo)

    # Monta o objeto final limpo
    novo_automato = {
        "alfabeto": alfabeto,
        "estados": novos_estados,
        "estados_iniciais": novos_iniciais,
        "estados_finais": novos_finais,
        "transicoes": novas_transicoes
    }

    return novo_automato

File: pascal/conversoes/afn_para_afd.pas
unit afn_para_afd;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, af_utils;

function Converter(const Automato: TAutomato): TAutomato;

implementation

function ConjuntoParaNome(const Conjunto: TStringList): string;
begin
  Result := NormalizarNomeEstado(Conjunto);
end;

function ClonarConjunto(const Fonte: TStringList): TStringList;
begin
  Result := TStringList.Create;
  Result.Sorted := True;
  Result.Duplicates := dupIgnore;
  Result.Assign(Fonte);
end;

function ConjuntoExiste(const Lista: array of TStringList; const Cmp: TStringList): Integer;
var
  I: Integer;
begin
  for I := 0 to Length(Lista) - 1 do
    if Lista[I].Equals(Cmp) then
      Exit(I);
  Result := -1;
end;

function Converter(const Automato: TAutomato): TAutomato;
var
  Alfabeto: TStringArray;
  EstadosAFN, EstadosFinaisAFN: TStringList;
  EstadoInicialAFN: TStringList;
  EstadosBrutos: TStringList;
  Visitados: array of TStringList;
  Fila: array of Integer;
  FilaInicio, FilaFim: Integer;
  TransicoesBrutas: TTransicoes;
  EstadoAtualIdx: Integer;
  EstadoAtualNome, NomeDestino: string;
  Simbolo: string;
  Destino: TStringList;
  MapaNomes: TStringList;
  NovosEstados, NovosIniciais, NovosFinais: TStringList;
  I, Posicao: Integer;
  NovoNome: string;
  Prefixo: string;
  NovaTrans: TTransicao;
begin
  Alfabeto := Automato.Alfabeto;

  EstadosAFN := StringArrayToList(Automato.Estados);
  EstadosFinaisAFN := StringArrayToList(Automato.EstadosFinais);

  EstadoInicialAFN := StringArrayToList(Automato.EstadosIniciais);

  EstadosBrutos := TStringList.Create;
  EstadosBrutos.Sorted := False;
  EstadosBrutos.Duplicates := dupIgnore;

  SetLength(Visitados, 1);
  Visitados[0] := ClonarConjunto(EstadoInicialAFN);
  EstadosBrutos.Add(ConjuntoParaNome(Visitados[0]));

  SetLength(Fila, Length(Fila) + 1);
  Fila[0] := 0;
  FilaInicio := 0;
  FilaFim := 1;

  SetLength(TransicoesBrutas, 0);

  while FilaInicio < FilaFim do
  begin
    EstadoAtualIdx := Fila[FilaInicio];
    Inc(FilaInicio);
    EstadoAtualNome := ConjuntoParaNome(Visitados[EstadoAtualIdx]);

    for Simbolo in Alfabeto do
    begin
      Destino := Mover(Automato, Visitados[EstadoAtualIdx], Simbolo);
      NomeDestino := ConjuntoParaNome(Destino);

      SetLength(TransicoesBrutas, Length(TransicoesBrutas) + 1);
      TransicoesBrutas[High(TransicoesBrutas)].Origem := EstadoAtualNome;
      TransicoesBrutas[High(TransicoesBrutas)].Destino := NomeDestino;
      TransicoesBrutas[High(TransicoesBrutas)].Simbolo := Simbolo;

      Posicao := ConjuntoExiste(Visitados, Destino);
      if Posicao = -1 then
      begin
        SetLength(Visitados, Length(Visitados) + 1);
        Visitados[High(Visitados)] := Destino;
        EstadosBrutos.Add(NomeDestino);

        SetLength(Fila, Length(Fila) + 1);
        Fila[FilaFim] := High(Visitados);
        Inc(FilaFim);
      end
      else
        Destino.Free;
    end;
  end;

  MapaNomes := TStringList.Create;
  MapaNomes.Sorted := False;
  MapaNomes.Duplicates := dupIgnore;

  NovosEstados := TStringList.Create;
  NovosEstados.Sorted := False;
  NovosEstados.Duplicates := dupIgnore;

  NovosIniciais := TStringList.Create;
  NovosIniciais.Sorted := False;
  NovosIniciais.Duplicates := dupIgnore;

  NovosFinais := TStringList.Create;
  NovosFinais.Sorted := False;
  NovosFinais.Duplicates := dupIgnore;

  Prefixo := 'S';
  for I := 0 to EstadosBrutos.Count - 1 do
  begin
    if EstadosBrutos[I] = '{}' then
      NovoNome := 'q_erro'
    else
      NovoNome := Prefixo + IntToStr(I);

    MapaNomes.Add(EstadosBrutos[I] + '=' + NovoNome);
    NovosEstados.Add(NovoNome);
  end;

  for I := 0 to Length(TransicoesBrutas) - 1 do
  begin
    NovoNome := MapaNomes.Values[TransicoesBrutas[I].Origem];
    if NovoNome = '' then
      NovoNome := TransicoesBrutas[I].Origem;
    TransicoesBrutas[I].Origem := NovoNome;

    NovoNome := MapaNomes.Values[TransicoesBrutas[I].Destino];
    if NovoNome = '' then
      NovoNome := TransicoesBrutas[I].Destino;
    TransicoesBrutas[I].Destino := NovoNome;
  end;

  for I := 0 to EstadosBrutos.Count - 1 do
  begin
    if I = 0 then
      NovosIniciais.Add(MapaNomes.Values[EstadosBrutos[I]]);

    if EstadosFinaisCompostos(Visitados[I], EstadosFinaisAFN) then
      NovosFinais.Add(MapaNomes.Values[EstadosBrutos[I]]);
  end;

  Result.Alfabeto := Alfabeto;
  Result.Estados := ListToStringArray(NovosEstados);
  Result.EstadosIniciais := ListToStringArray(NovosIniciais);
  Result.EstadosFinais := ListToStringArray(NovosFinais);
  Result.Transicoes := TransicoesBrutas;

  EstadosAFN.Free;
  EstadosFinaisAFN.Free;
  EstadoInicialAFN.Free;
  EstadosBrutos.Free;
  for I := 0 to Length(Visitados) - 1 do
    Visitados[I].Free;
  MapaNomes.Free;
  NovosEstados.Free;
  NovosIniciais.Free;
  NovosFinais.Free;
end;

end.
